### PR TITLE
HDDS-9806. Refactor TestBlockOutputStream

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Reduce code duplication:
* `TestBlockOutputStream` is duplicated as `TestBlockOutputStreamFlushDelay`
* `TestBlockOutputStreamWithFailures` is duplicated as `TestBlockOutputStreamWithFailuresFlushDelay`

The main difference between the duplicates is that the ones without `FlushDelay` set `OzoneClientConfig.setStreamBufferFlushDelay(false)`, while the ones with `FlushDelay` rely on the default value for this property being `true`.

Furthermore, `streamBufferFlushDelay` is a client-side config, so we can completely merge `TestBlockOutputStream` and `TestBlockOutputStreamFlushDelay` by:
 * making test cases parameterized
 * creating a new client for each test case

https://issues.apache.org/jira/browse/HDDS-9806

## How was this patch tested?

CI:
https://github.com/adoroszlai/ozone/actions/runs/7047698557

### CODE REVIEW

**Comment by kerneltime** [comment]
@DaveTeng0 can you please take a look?

**Comment by adoroszlai** [comment]
@ArafatKhan2198 @devmadhuu please review

**Review by devmadhuu — APPROVED** [review]
Thanks Attila for working on this patch. LGTM +1

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestBlockOutputStream.java [hunk 1]
```diff
@@ -18,691 +18,690 @@
...
-
-    // make sure the bufferPool is empty
-    Assert
-        .assertEquals(0, blockOutputStream.getBufferPool().computeBufferData());
-    Assert.assertEquals(dataLength, blockOutputStream.getTotalAckDataLength());
-    Assert.assertTrue(
-        blockOutputStream.getCommitIndex2flushedDataMap().isEmpty());
-    Assert.assertEquals(0, keyOutputStream.getStreamEntries().size());
-    validateData(keyName, data1);
+  static OzoneClientConfig newClientConfig(ConfigurationSource source,
```
Comment (hemantk-12) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestBlockOutputStream.java:143]: nit:  Isn't it supposed to be like this, according to ozone style?
```suggestion
  static OzoneClientConfig newClientConfig(ConfigurationSource source,
                                           boolean flushDelay) {
```

**Review by hemantk-12 — APPROVED** [review]
LGTM.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestBlockOutputStream.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestBlockOutputStream.java:143]: Aligning parameters this way causes unnecessary extra reformatting if:

 * method is renamed
 * return type is renamed or changed
 * visibility is changed

Comment (hemantk-12) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/TestBlockOutputStream.java:143]: That's a fair point.
(Not for this PR, we should update [ozone-style](https://github.com/apache/ozone/blob/master/hadoop-ozone/dev-support/intellij/ozone-style.xml) to align parameters in same way as you suggested in that case.)

**Comment by adoroszlai** [comment]
Thanks @ArafatKhan2198, @DaveTeng0, @devmadhuu, @hemantk-12 for the review.

**Pull request merged**